mrp_production_batch/models/mrp_production_batch.py

```diff
@@ -0,0 +1,190 @@
+        for rec in self:
+            rec.product_qty = sum(rec.production_ids.mapped("product_qty"))
+
+    product_ids = fields.Many2many(
+        comodel_name="product.product",
+        compute="_compute_product_ids",
+        string="Products",
+    )
+
+    def _compute_product_ids(self):
+        for rec in self:
+            rec.product_ids = rec.production_ids.mapped("product_id")
+
+    def action_done(self):
+        self.ensure_one()
+        productions = self.production_ids.filtered(lambda r: r.state != "done")
+        len_production = len(productions)
+        productions_display_name = []
+        for production in productions:
+            try:
+                production.with_context(
+                    mrp_production_batch_create=True
+                ).button_mark_done()
+            except Exception:
+                productions_display_name.append(production.display_name)
+        if productions_display_name and len(productions_display_name) == len_production:
+            message = "The following productions could not be marked as 'done':\n"
+            message += "\n".join(productions_display_name)
+            raise UserError(message)
+
+    def _get_common_action_view_production(self):
+        tree_view = self.env.ref("mrp.mrp_production_tree_view")
+        form_view = self.env.ref("mrp.mrp_production_form_view")
+        return {
+            "name": _("Detailed Operations"),
+            "type": "ir.actions.act_window",
+            "view_mode": "tree,form",
+            "res_model": "mrp.production",
+            "views": [(tree_view.id, "tree"), (form_view.id, "form")],
+            "view_id": tree_view.id,
+        }
+
+    def action_view_total_production(self):
+        self.ensure_one()
+        action = self._get_common_action_view_production()
+        action["domain"] = [("id", "in", self.production_ids.ids)]
+        return action
+
+    def action_view_production_pending(self):
+        self.ensure_one()
+        action = self._get_common_action_view_production()
+        action["domain"] = [
+            ("id", "in", self.production_ids.ids),
+            ("state", "not in", ("done", "cancel")),
+        ]
+        return action
+
+    def action_view_production_done(self):
+        self.ensure_one()
+        action = self._get_common_action_view_production()
+        action["domain"] = [
+            ("id", "in", self.production_ids.ids),
+            ("state", "=", "done"),
+        ]
+        return action
+
+    def action_view_production_cancelled(self):
+        self.ensure_one()
+        action = self._get_common_action_view_production()
+        action["domain"] = [
+            ("id", "in", self.production_ids.ids),
+            ("state", "=", "cancel"),
+        ]
+        return action
+
+    def action_generate_serial(self):
+        for rec in self:
+            for production in rec.production_wo_lot_producing_id:
+                production.action_generate_serial()
```

Comment:
We need to change this, the serial numbers should be generated at the time the MO is processed.